File: config/cache.py
"""
Configuration cache helpers used to persist harvested proxies.
"""
from typing import Dict, List, Optional, Set

from hunter.core.utils import append_unique_lines, load_json, read_lines, save_json, write_lines
from hunter.core.utils import now_ts


class SmartCache:
    """Smart caching with failure tracking."""

    def __init__(self, cache_file: str = "subscriptions_cache.txt", working_cache_file: str = "working_configs_cache.txt") -> None:
        self.cache_file = cache_file
        self.working_cache_file = working_cache_file
        self._last_successful_fetch = 0
        self._consecutive_failures = 0

    def save_configs(self, configs: List[str], working: bool = False) -> int:
        target = self.working_cache_file if working else self.cache_file
        appended = append_unique_lines(target, configs)
        if appended > 0:
            self._last_successful_fetch = now_ts()
            self._consecutive_failures = 0
        return appended

    def load_cached_configs(self, max_count: int = 1000, working_only: bool = False) -> Set[str]:
        configs = set()
        sources = [self.cache_file]
        if working_only:
            sources = [self.working_cache_file]
        else:
            sources.append(self.working_cache_file)
        for path in sources:
            lines = read_lines(path)
            for line in lines[-max_count:]:
                if line and "://" in line:
                    configs.add(line)
        return configs

    def record_failure(self) -> None:
        self._consecutive_failures += 1

    def should_use_cache(self) -> bool:
        return self._consecutive_failures >= 2

    def get_failure_count(self) -> int:
        return self._consecutive_failures


class ResilientHeartbeat:
    """Heartbeat monitor for Telegram clients."""

    def __init__(self, check_interval: int = 30) -> None:
        self.check_interval = check_interval
        self._last_heartbeat = now_ts()
        self._is_connected = False
        self._reconnect_attempts = 0
        self._max_reconnect_attempts = 5

    async def check_connection(self, client: Optional["TelegramClient"]) -> bool:
        if client is None:
            self._is_connected = False
            return False
        try:
            if not client.is_connected():
                self._is_connected = False
                return False
            await client.get_me()
            self._is_connected = True
            self._last_heartbeat = now_ts()
            self._reconnect_attempts = 0
            return True
        except Exception:
            self._is_connected = False
            return False

    async def try_reconnect(self, client: Optional["TelegramClient"]) -> bool:
        if self._reconnect_attempts >= self._max_reconnect_attempts:
            return False
        self._reconnect_attempts += 1
        try:
            if client:
                await client.disconnect()
                await client.connect()
                if await client.is_user_authorized():
                    self._is_connected = True
                    return True
        except Exception:
            return False
        return False

    def is_connected(self) -> bool:
        return self._is_connected

    def time_since_heartbeat(self) -> int:
        return now_ts() - self._last_heartbeat

File: native/android/app/src/main/cpp/cache/cache.cpp
#include "cache.h"
#include "core/utils.h"

namespace hunter {

// ---------- SmartCache ----------

SmartCache::SmartCache(const std::string& cache_file, const std::string& working_cache_file)
    : cache_file_(cache_file), working_cache_file_(working_cache_file) {}

void SmartCache::set_base_dir(const std::string& dir) {
    base_dir_ = dir;
}

std::string SmartCache::resolve_path(const std::string& filename) const {
    if (base_dir_.empty()) return filename;
    return base_dir_ + "/" + filename;
}

int SmartCache::save_configs(const std::vector<std::string>& configs, bool working) {
    std::string target = working ? resolve_path(working_cache_file_) : resolve_path(cache_file_);
    int appended = append_unique_lines(target, configs);
    if (appended > 0) {
        last_successful_fetch_ = now_ts();
        consecutive_failures_ = 0;
    }
    return appended;
}

std::set<std::string> SmartCache::load_cached_configs(int max_count, bool working_only) {
    std::set<std::string> configs;
    std::vector<std::string> sources;

    if (working_only) {
        sources.push_back(resolve_path(working_cache_file_));
    } else {
        sources.push_back(resolve_path(cache_file_));
        sources.push_back(resolve_path(working_cache_file_));
    }

    for (const auto& path : sources) {
        auto lines = read_lines(path);
        // Take last max_count lines
        size_t start = (lines.size() > static_cast<size_t>(max_count))
                       ? lines.size() - max_count : 0;
        for (size_t i = start; i < lines.size(); ++i) {
            if (!lines[i].empty() && lines[i].find("://") != std::string::npos) {
                configs.insert(lines[i]);
            }
        }
    }

    return configs;
}

void SmartCache::record_failure() {
    consecutive_failures_++;
}

bool SmartCache::should_use_cache() const {
    return consecutive_failures_ >= 2;
}

int SmartCache::get_failure_count() const {
    return consecutive_failures_;
}

// ---------- ResilientHeartbeat ----------

ResilientHeartbeat::ResilientHeartbeat()
    : last_heartbeat_(now_ts()) {}

bool ResilientHeartbeat::is_connected() const {
    return is_connected_;
}

int64_t ResilientHeartbeat::time_since_heartbeat() const {
    return now_ts() - last_heartbeat_;
}

void ResilientHeartbeat::mark_connected() {
    is_connected_ = true;
    last_heartbeat_ = now_ts();
    reconnect_attempts_ = 0;
}

void ResilientHeartbeat::mark_disconnected() {
    is_connected_ = false;
}

bool ResilientHeartbeat::should_attempt_reconnect() {
    if (reconnect_attempts_ >= max_reconnect_attempts_) return false;
    reconnect_attempts_++;
    return true;
}

void ResilientHeartbeat::reset() {
    is_connected_ = false;
    reconnect_attempts_ = 0;
    last_heartbeat_ = now_ts();
}

} // namespace hunter
